File: shadowdroid-plugin/src/main/java/io/github/andriyo/shadowdroid/studio/SourceResolver.java
package io.github.andriyo.shadowdroid.studio;

import com.intellij.openapi.diagnostic.Logger;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

record SourceResolver(Map<String, List<String>> byName) {
    private static final Logger LOG = Logger.getInstance(SourceResolver.class);

    static SourceResolver build(String basePath) {
        if (basePath == null || basePath.isBlank()) {
            return new SourceResolver(Collections.emptyMap());
        }
        Map<String, List<String>> byName = new LinkedHashMap<>();
        Path root = Path.of(basePath);
        try (var paths = Files.walk(root)) {
            paths
                .filter(Files::isRegularFile)
                .forEach(path -> {
                    String normalized = path.toAbsolutePath().toString().replace('\\', '/');
                    if (!isSourcePath(normalized)) return;
                    String name = path.getFileName() == null ? "" : path.getFileName().toString();
                    if (name.isBlank()) return;
                    byName.computeIfAbsent(name, ignored -> new ArrayList<>()).add(path.toAbsolutePath().toString());
                });
        } catch (Throwable t) {
            LOG.debug("Unable to build source resolver for " + basePath, t);
        }
        for (List<String> paths : byName.values()) {
            Collections.sort(paths);
        }
        return new SourceResolver(byName);
    }

    String resolve(String composeFilename) {
        if (composeFilename == null || composeFilename.isBlank()) return null;
        String normalized = composeFilename.replace('\\', '/');
        String fileName = normalized.substring(normalized.lastIndexOf('/') + 1);
        if (fileName.isBlank()) return null;
        List<String> candidates = byName.get(fileName);
        if (candidates == null || candidates.isEmpty()) return null;
        String fallback = null;
        for (String candidate : candidates) {
            String path = candidate.replace('\\', '/');
            if (fallback == null) fallback = candidate;
            if (path.endsWith(normalized)) return candidate;
        }
        return fallback;
    }

    private static boolean isSourcePath(String path) {
        if (path.contains("/build/") || path.contains("/.gradle/") || path.contains("/.idea/")) return false;
        return path.endsWith(".kt") || path.endsWith(".java") || path.endsWith(".xml");
    }
}

File: shadowdroid-plugin/src/main/kotlin/io/github/andriyo/shadowdroid/studio/SourceResolver.kt
package io.github.andriyo.shadowdroid.studio

import com.intellij.openapi.diagnostic.Logger
import java.io.IOException
import java.nio.file.FileVisitResult
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.SimpleFileVisitor
import java.nio.file.attribute.BasicFileAttributes

internal data class SourceResolver(
    private val byName: Map<String, List<String>>,
) {
    fun resolve(composeFilename: String?): String? {
        if (composeFilename.isNullOrBlank()) return null
        val normalized = composeFilename.replace('\\', '/')
        val fileName = normalized.substringAfterLast('/')
        if (fileName.isBlank()) return null
        val candidates = byName[fileName].orEmpty()
        if (candidates.isEmpty()) return null
        return candidates.firstOrNull { it.replace('\\', '/').endsWith(normalized) } ?: candidates.first()
    }

    companion object {
        private val LOG = Logger.getInstance(SourceResolver::class.java)
        private val SKIPPED_DIRECTORIES = setOf("build", ".gradle", ".idea", ".git", "node_modules")

        @JvmStatic
        fun build(basePath: String?): SourceResolver {
            if (basePath.isNullOrBlank()) return SourceResolver(emptyMap())
            val byName = linkedMapOf<String, MutableList<String>>()
            val root = Path.of(basePath)
            try {
                Files.walkFileTree(root, object : SimpleFileVisitor<Path>() {
                    override fun preVisitDirectory(dir: Path, attrs: BasicFileAttributes): FileVisitResult =
                        if (dir.fileName?.toString() in SKIPPED_DIRECTORIES) FileVisitResult.SKIP_SUBTREE
                        else FileVisitResult.CONTINUE

                    override fun visitFile(file: Path, attrs: BasicFileAttributes): FileVisitResult {
                        if (attrs.isRegularFile) {
                            val name = file.fileName?.toString().orEmpty()
                            if (name.isNotBlank() && isSourceFile(name)) {
                                byName.getOrPut(name) { mutableListOf() }.add(file.toAbsolutePath().toString())
                            }
                        }
                        return FileVisitResult.CONTINUE
                    }

                    override fun visitFileFailed(file: Path, exc: IOException): FileVisitResult =
                        FileVisitResult.CONTINUE
                })
            } catch (t: Throwable) {
                LOG.debug("Unable to build source resolver for $basePath", t)
            }
            byName.values.forEach { it.sort() }
            return SourceResolver(byName)
        }

        private fun isSourceFile(name: String): Boolean =
            name.endsWith(".kt") || name.endsWith(".java") || name.endsWith(".xml")
    }
}
